### PR TITLE
Simulation mem leak - create hard block mappings

### PR DESCRIPTION
**Pull request by sdamghan**
#### How Has This Been Tested?
tested by make test odin

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: ODIN_II/SRC/read_blif.cpp [hunk 1]
```diff
@@ -640,6 +640,9 @@ void create_hard_block_nodes(hard_block_models *models, FILE *file, Hashtable *o
   	free_hard_block_ports(ports);
   	mapping_index->destroy_free_items();
 	delete mapping_index;
+	for (i = 0; i < count; i++)
+		vtr::free(mappings[i]);
+
```
Comment (jeanlego) [inline ODIN_II/SRC/read_blif.cpp]: move up to line 536

Comment (sdamghan) [reply ODIN_II/SRC/read_blif.cpp]: the line 540 uses mapping array for "get_hard_block_ports", we can't free them before this line.

Comment (jeanlego) [reply ODIN_II/SRC/read_blif.cpp]: get_hard_block_ports is at line 535, hence why i think the freeing should be moved ate 536

File: ODIN_II/SRC/read_blif.cpp [hunk 2]
```diff
@@ -534,6 +534,9 @@ void create_hard_block_nodes(hard_block_models *models, FILE *file, Hashtable *o
 	// Index the mappings in a hard_block_ports struct.
 	hard_block_ports *ports = get_hard_block_ports(mappings, count);
 
+	for (i = 0; i < count; i++)
+		vtr::free(mappings[i]);
+
```
Comment (jeanlego) [inline ODIN_II/SRC/read_blif.cpp]: make sure you set mappings[i] = NULL in the for loop.

Comment (jeanlego) [reply ODIN_II/SRC/read_blif.cpp]: also free the mappings array here and set to NULL

File: ODIN_II/SRC/read_blif.cpp (hunk 1 above)
Comment (jeanlego) [reply ODIN_II/SRC/read_blif.cpp]: once moved, please remove all the occurence of mappings down here

File: ODIN_II/SRC/read_blif.cpp [hunk 3]
```diff
@@ -535,7 +535,15 @@ void create_hard_block_nodes(hard_block_models *models, FILE *file, Hashtable *o
 	hard_block_ports *ports = get_hard_block_ports(mappings, count);
 
 	for (i = 0; i < count; i++)
+	{
+		if (mappings[i] != NULL)
```
Comment (jeanlego) [inline ODIN_II/SRC/read_blif.cpp]: you just lost your pointer there.

set to NULL after free,
don't check if its NULL before setting to NULL, its a bit redundant

File: ODIN_II/SRC/read_blif.cpp [hunk 4]
```diff
@@ -535,7 +535,15 @@ void create_hard_block_nodes(hard_block_models *models, FILE *file, Hashtable *o
...
+			mappings[i] = NULL;
+
 		vtr::free(mappings[i]);
+	}
+	
+	mappings = NULL;
```
Comment (jeanlego) [inline ODIN_II/SRC/read_blif.cpp]: same here

File: ODIN_II/SRC/read_blif.cpp (hunk 3 above)
Comment (jeanlego) [reply ODIN_II/SRC/read_blif.cpp]: https://www.eskimo.com/~scs/cclass/notes/sx11b.html

File: ODIN_II/SRC/read_blif.cpp [hunk 5]
```diff
@@ -640,7 +650,9 @@ void create_hard_block_nodes(hard_block_models *models, FILE *file, Hashtable *o
   	free_hard_block_ports(ports);
   	mapping_index->destroy_free_items();
 	delete mapping_index;
-  	vtr::free(mappings);
+	for (i = 0; i < count; i++)
```
Comment (jeanlego) [inline ODIN_II/SRC/read_blif.cpp]: Remove these line (mapping is set to NULL, you can't access anything inside it. This is why I've made you free above and set to NULL)

**Pull request merged**